File: components/python/src/kitchen_manager.py
import asyncio
from typing import Any, Set

from fastapi import WebSocket


class KitchenManager:
    def __init__(self):
        self._clients: Set[WebSocket] = set()
        self._lock = asyncio.Lock()

    async def register(self, ws: WebSocket) -> None:
        async with self._lock:
            self._clients.add(ws)

    async def unregister(self, ws: WebSocket) -> None:
        async with self._lock:
            self._clients.discard(ws)

    async def broadcast(self, message: dict[str, Any]) -> None:
        dead: set[WebSocket] = set()
        async with self._lock:
            for ws in self._clients:
                try:
                    await ws.send_json(message)
                except Exception:
                    dead.add(ws)
            for ws in dead:
                self._clients.discard(ws)


kitchen_manager = KitchenManager()

File: components/typescript/src/kitchen-manager.ts
import type { WSContext } from "hono/ws";
import type WebSocket from "ws";

export interface KitchenOrder {
  id: string;
  items: {
    product_id: string;
    name: string;
    quantity: number;
    unit_price_cents: number;
  }[];
  total_cents: number;
  status: "nuevo" | "en preparación" | "listo";
  created_at: string;
}

export type KitchenServerEvent =
  | { type: "new_order"; order: KitchenOrder }
  | {
      type: "status_updated";
      orderId: string;
      status: KitchenOrder["status"];
    }
  | { type: "orders_snapshot"; orders: KitchenOrder[] };

export type KitchenClientEvent = {
  type: "update_status";
  orderId: string;
  status: KitchenOrder["status"];
};

export class KitchenManager {
  private clients = new Set<WSContext<WebSocket>>();
  private orders = new Map<string, KitchenOrder>();

  register(ws: WSContext<WebSocket>): void {
    this.clients.add(ws);
    this.sendToClient(ws, {
      type: "orders_snapshot",
      orders: Array.from(this.orders.values()),
    });
  }

  unregister(ws: WSContext<WebSocket>): void {
    this.clients.delete(ws);
  }

  addOrder(order: KitchenOrder): void {
    this.orders.set(order.id, order);
    this.broadcast({ type: "new_order", order });
  }

  updateStatus(
    orderId: string,
    status: KitchenOrder["status"]
  ): boolean {
    const order = this.orders.get(orderId);
    if (!order) return false;
    order.status = status;
    this.broadcast({ type: "status_updated", orderId, status });
    return true;
  }

  getOrders(): KitchenOrder[] {
    return Array.from(this.orders.values());
  }

  broadcast(event: KitchenServerEvent): void {
    const msg = JSON.stringify(event);
    for (const ws of this.clients) {
      try {
        ws.send(msg);
      } catch {
        this.clients.delete(ws);
      }
    }
  }

  private sendToClient(
    ws: WSContext<WebSocket>,
    event: KitchenServerEvent
  ): void {
    try {
      ws.send(JSON.stringify(event));
    } catch {
      this.clients.delete(ws);
    }
  }
}

export const kitchenManager = new KitchenManager();
